### PR TITLE
Nameservice interact

### LINKED ISSUE #157: make script for interacting with the nameservice app
this script should run in a loop (indefinitely) simulating interacting with the nameservice app using all message types

### PR DESCRIPTION
**Pull request by IvantheTricourne**
Closes #157

### CODE REVIEW

**Comment by IvantheTricourne** [comment]
Merge after #151 and #162

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+user1 = makeUser "f65255094d7773ed8dd417badc9fc045c1f80fdc5b2d25172b031ce6933e039a"
+
+user2 :: User
+user2 = makeUser "f65242094d7773ed8dd417badc9fc045c1f80fdc5b2d25172b031ce6933e039a"
+
+--------------------------------------------------------------------------------
+-- Actions
+--------------------------------------------------------------------------------
+
+genNames :: Int -> IO [Name]
```
Comment (blinky3713) [inline hs-abci-examples/nameservice/interact/Interact.hs]: genNames = replicateM (fmap fromString genName)

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+genNames 0 = return []
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Interact.hs:42]: genNames = replicateM (fmap fromString genName)

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+genNames x = do
+  genName <- Name.name
+  let aName = fromString genName
+  names <- genNames (x - 1)
+  return (aName:names)
+
+genVals :: Int -> IO [Text]
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Interact.hs:49]: genVals = replicateM (fmap cs Lorem.word)

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+genVals 0 = return []
+genVals x = do
+  genVal <- Lorem.word
+  vals <- genVals (x - 1)
+  return (cs genVal:vals)
+
+createName :: User -> Name -> Text -> IO ()
+createName user name val = buyName user name val 0
+
+createNames :: User -> [(Name, Text)] -> IO ()
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Interact.hs:59]: createNames user = mapM_ (uncurry . createName user)

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  createName user name val
+  createNames user rst
+
+buyName :: User -> Name -> Text -> Amount -> IO ()
+buyName User{userAddress, userPrivKey} name newVal amount =
+  let msg = TypedMessage "BuyName" (encode $ BuyName amount name newVal userAddress)
+      rawTx = mkSignedRawTransactionWithRoute "nameservice" userPrivKey msg
+  in runTransaction_ rawTx
+
+buyNames :: User -> [(Name, Text, Amount)] -> IO ()
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Interact.hs:71]: buyNames user = mapM_ (uncurry . buyName user)

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  buyName user name val amt
+  buyNames user rst
+
+deleteName :: User -> Name -> IO ()
+deleteName User{userAddress, userPrivKey} name =
+  let msg = TypedMessage "DeleteName" (encode $ DeleteName userAddress name)
+      rawTx = mkSignedRawTransactionWithRoute "nameservice" userPrivKey msg
+  in runTransaction_ rawTx
+
+deleteNames :: User -> [Name] -> IO ()
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Interact.hs:83]: createNames user = mapM_ (uncurry . deleteName user)

File: hs-abci-examples/nameservice/interact/Interact.hs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+setName :: User -> Name -> Text -> IO ()
+setName User{userAddress, userPrivKey} name val =
+  let msg = TypedMessage "SetName" (encode $ SetName name userAddress val)
+      rawTx = mkSignedRawTransactionWithRoute "nameservice" userPrivKey msg
+  in runTransaction_ rawTx
+
+setNames :: User -> [(Name, Text)] -> IO ()
+setNames _ [] = return ()
+setNames user ((name, val):rst) = do
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Interact.hs:97]: createNames user = mapM_ (uncurry . setName user)

File: hs-abci-examples/nameservice/interact/Main.hs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import           Interact
+import           System.Environment (lookupEnv)
+import           Text.Read          (read)
+
+main :: IO ()
+main = do
+  mConc <- lookupEnv "TX_COUNT"
+  let conc = maybe 1 read mConc
+  putStrLn $ "Running nameservice interaction w/ TX_COUNT: " <> show conc
+  faucetAccount user1 10000
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Main.hs:14]: you should faucet inside the loop

File: hs-abci-examples/nameservice/interact/Main.hs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+import           Control.Monad      (forever)
+import           Data.Maybe         (maybe)
+import           Interact
+import           System.Environment (lookupEnv)
+import           Text.Read          (read)
+
+main :: IO ()
+main = do
+  mConc <- lookupEnv "TX_COUNT"
```
Comment (martyall) [inline hs-abci-examples/nameservice/interact/Main.hs:11]: what is mConc? use a variable name that's more transparent

**Review by martyall — CHANGES_REQUESTED** [review]

**Pull request closed without merging**